Fix Source filtering in new field collapsing feature - Closes #24063

byronvoorbach opened the pull request:
Added 'OBJECT_ARRAY_BOOLEAN_OR_STRING' to the ObjectParser.ValueType enum to be in line with the current configuration for source filtering in the request body. Updated InnerHitBuilder to use the new enum type. InnerHitBuilder now supports Object, Array, String or Boolean values for _source filtering

Ref: https://github.com/elastic/elasticsearch/issues/24063

karmi commented:
Hi @byronvoorbach, we have found your signature in our records, but it seems like you have signed with a different e-mail than the one used in yout Git [commit](https://github.com/elastic/elasticsearch/pull/24068.patch). Can you please add both of these e-mails into your Github profile (they can be hidden), so we can match your e-mails to your Github profile?

elasticmachine commented:
Since this is a community submitted pull request, a Jenkins build has not been kicked off automatically. Can an Elastic organization member please verify the contents of this patch and then kick off a build manually?

byronvoorbach commented:
Whoops, that's what you get when working with multiple pc's... Added the other mail address!

olcbean commented:
@byronvoorbach Thank you for contributing 😃 
Could you elaborate what problem this PR is fixing?

nik9000 commented:
Thanks for fixing your email @byronvoorbach. More administrative stuff:
1. Can you change the PR's title to make it clear what you changed? Most of us use email to decide what to review and without a description of what is changed or fixed we aren't likely to pick the PR to review.
2. Can you add a description? Sort of for the same reason as the above point. If you add `Closes #24063` then we'll get the nice automation when we merge the fix.

jimczi commented on core/src/main/java/org/elasticsearch/index/query/InnerHitBuilder.java:
```diff
@@ -101,7 +101,7 @@
                 throw new ParsingException(p.getTokenLocation(), "Could not parse inner _source definition", e);
             }
-        }, SearchSourceBuilder._SOURCE_FIELD, ObjectParser.ValueType.OBJECT_OR_BOOLEAN);
+        }, SearchSourceBuilder._SOURCE_FIELD, ObjectParser.ValueType.OBJECT_ARRAY_OR_STRING);
```
Not a big deal but this would break the parsing as a boolean eg.`_source:true`. You can maybe add a OBJECT_ARRAY_BOOLEAN_OR_STRING to the enum ?

byronvoorbach replied:
Yes! Will swap it out for that one and test

jimczi commented on core/src/test/java/org/elasticsearch/index/query/InnerHitBuilderTests.java:
```diff
@@ -312,7 +313,7 @@ public static InnerHitBuilder randomInnerHits(boolean recursive, boolean include
         FetchSourceContext randomFetchSourceContext;
         if (randomBoolean()) {
-            randomFetchSourceContext = new FetchSourceContext(randomBoolean());
+            randomFetchSourceContext = new FetchSourceContext(true, Strings.EMPTY_ARRAY, Strings.EMPTY_ARRAY);
```
Can you add more randomization here. You can build a random list of field names for the includes and the excludes for instance.

byronvoorbach replied:
Thanks for the comments @jimczi, I'm away from home till after the Easter weekend. Will pick them up when back!

byronvoorbach commented:
@jimczi Sorry for the delay. I've been quite busy at one of my clients. I updated the tests and added OBJECT_ARRAY_BOOLEAN_OR_STRING. The test-cases from my gist seem to pass now. Any more comments? :)

elasticmachine commented:
Since this is a community submitted pull request, a Jenkins build has not been kicked off automatically. Can an Elastic organization member please verify the contents of this patch and then kick off a build manually?

jimczi commented:
retest this please

byronvoorbach commented:
@jimczi Cool! Thanks voor the headsup!!

byronvoorbach commented:
Hey @jimczi, I noticed that the build is failing. Is there something you need from me?

jimczi commented:
@byronvoorbach the build is failing but it is not related to this PR. I am currently testing your PR locally. I'll merge if the tests pass on my machine

jimczi commented:
Thanks @byronvoorbach 
The fix will be available in 5.5.0

elasticmachine commented:
Since this is a community submitted pull request, a Jenkins build has not been kicked off automatically. Can an Elastic organization member please verify the contents of this patch and then kick off a build manually?
